### PR TITLE
Add string ctor to MemberNotNull/When

### PR DESCRIPTION
**Pull request by jcouv**
Follow-up on PR https://github.com/dotnet/roslyn/pull/41675 (Add support for MemberNotNull/When attributes)

The constructors with `string[]` are not CLS-compliant and that causes problems for the BCL team. From email discussion with LDM, we'll add simple constructors that take a single member name at a time, and we'll allow the attributes to be instantiated multiple times.

Relates to https://github.com/dotnet/roslyn/issues/41438 (Implement `MemberNotNull` nullability attribute)
Corresponding BCL change: https://github.com/dotnet/runtime/pull/33864

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Attributes/SourceAttributeData.cs [hunk 1]
```diff
@@ -269,116 +276,155 @@ internal override int GetTargetAttributeSignatureIndex(Symbol targetSymbol, Attr
                             specType = parameterType.GetEnumUnderlyingType()!.SpecialType;
                         }
                     }
-                    else if (parameterType.IsArray())
+                    else if (targetType != (byte)SignatureTypeCode.SZArray && parameterType.IsArray())
                     {
+                        if (targetSignature[signatureByteIndex - 1] != (byte)SignatureTypeCode.SZArray)
+                        {
+                            return false;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Attributes/SourceAttributeData.cs]: :memo: there was a bug here. This method could not distinguish properly between the `string` case and the `string[]` case. To tighten that logic, I had to refactor a bit (extract a local function).

File: src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs [hunk 2]
```diff
@@ -256,25 +256,37 @@ static internal void DecodeSkipLocalsInitAttribute<T>(CSharpCompilation compilat
...
             {
                 return;
             }
 
-            var builder = ArrayBuilder<string>.GetInstance();
-            foreach (var member in membersArray.Values)
+            if (value.Kind != TypedConstantKind.Array)
             {
-                var memberName = member.DecodeValue<string>(SpecialType.System_String);
+                string? memberName = value.DecodeValue<string>(SpecialType.System_String);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs]: Won't this throw in the case where the customer uses a version of `MemberNotNull` where there is a ctor with a non-string argument? I'm unsure what our tolerance here is for malformed attributes. Know @RikkiGibson is hitting this with the obsolete change.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs]: We should get here in such case. This method is only called once we've found the attribute we're interested in by AttributeDescription (which verifies one of two signatures we want was matched)

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 3]
```diff
@@ -1220,34 +1233,38 @@ internal ImmutableArray<string> GetConditionalAttributeValues(EntityHandle token
...
+            var result = ArrayBuilder<string>.GetInstance(attrInfos.Count);
 
-            ArrayBuilder<string> extractMemberNotNullData(List<AttributeInfo> attrInfos)
+            foreach (var ai in attrInfos)
             {
-                if (attrInfos is null)
+                if (ai.SignatureIndex == 0)
                 {
-                    return null;
+                    if (TryExtractStringValueFromAttribute(ai.Handle, out var extracted))
```
Comment (gafter) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: >var [](start = 74, length = 3)

Please be explicit here: `string`

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 4]
```diff
@@ -1220,34 +1233,38 @@ internal ImmutableArray<string> GetConditionalAttributeValues(EntityHandle token
...
+                        {
+                            result.Add(extracted);
+                        }
+                    }
                 }
-
-                var result = ArrayBuilder<string>.GetInstance(attrInfos.Count);
-
-                foreach (var ai in attrInfos)
+                else if (TryExtractStringArrayValueFromAttribute(ai.Handle, out var extracted2))
```
Comment (gafter) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: >var [](start = 80, length = 3)

Same here `string[]`

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -23380,15 +23436,62 @@ public C()
...
+                //         get => true;
+                Diagnostic(ErrorCode.WRN_MemberNotNullWhen, "true").WithArguments("field1", "True").WithLocation(29, 16),
+                // (29,16): warning CS8775: Member 'field2' may not have a null value when exiting with `True`.
+                //         get => true;
+                Diagnostic(ErrorCode.WRN_MemberNotNullWhen, "true").WithArguments("field2", "True").WithLocation(29, 16),
+                // (29,16): warning CS8775: Member 'field3' may not have a null value when exiting with `True`.
+                //         get => true;
+                Diagnostic(ErrorCode.WRN_MemberNotNullWhen, "true").WithArguments("field3", "True").WithLocation(29, 16)
                 );
         }
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >} [](start = 8, length = 1)

Please add a test (if there isn't one already) with multiple applications of the params version of `MemberNotNull`.

Please add a test with `[MemberNotNull(member: null)]` and `[MemberNotNull(members: null)]` including repetitions.

Similarly for `MemberNotNullWhen`.

**Comment by gafter** [comment]
Looks good (Iteration 1) but please add a couple of tests as noted.

**Pull request merged**